pacman-helper.sh

```diff
@@ -212,88 +224,134 @@ quick_add () { # <file>...
+					test -n "$arch" ||
+					die "Could not determine architecture of '$path'"
+				}
+				key=${arch}_msys
+				;;
+			*.src.tar.gz|*.src.tar.xz|*.src.tar.zst)
+				arch=sources
+				key=
+				;;
+			*.sig)
+				# skip explicit signatures; we copy them automatically
+				continue
+				;;
+			*)
+				echo "Skipping unknown file: $file" >&2
+				continue
+				;;
+			esac
+			test -n "$arch" || die "Could not determine architecture for $path"
+			case " $architectures sources " in
+			*" $arch "*) ;;  # okay
+			*) echo "Skipping file with unknown arch: $file" >&2; continue;;
+			esac
 
-		echo "Copying $file to $arch/..." >&2
-		test -z "$key" || eval "$key=\$$key\\ $file"
-		all_files="$all_files $arch/$file"
+			echo "Copying $file to $arch/..." >&2
+			test -z "$key" || eval "$key=\$$key\\ $file"
+			all_files="$all_files $arch/$file"
 
-		if test ! -d "$dir/$arch"
-		then
-			git -C "$dir" rev-parse --quiet --verify refs/remotes/origin/$arch >/dev/null ||
-			git -C "$dir" fetch --depth=1 origin x86_64 aarch64 i686 ||
-			die "$dir: could not fetch from pacman-repo"
+			if test ! -d "$dir/$arch"
+			then
+				git -C "$dir" rev-parse --quiet --verify refs/remotes/origin/$arch >/dev/null ||
+				git -C "$dir" fetch --depth=1 origin x86_64 aarch64 i686 ||
+				die "$dir: could not fetch from pacman-repo"
 
-			git -C "$dir" worktree add -b $arch $arch origin/$arch ||
-			die "Could not initialize $dir/$arch"
-		fi
+				git -C "$dir" worktree add -b $arch $arch origin/$arch ||
+				die "Could not initialize $dir/$arch"
+			fi
 
-		cp "$path" "$dir/$arch" ||
-		die "Could not copy $path to $dir/$arch"
+			cp "$path" "$dir/$arch" ||
+			die "Could not copy $path to $dir/$arch"
 
-		if test -f "$path".sig
-		then
-			cp "$path".sig "$dir/$arch/" ||
-			die "Could not copy $path.sig to $dir/$arch"
-                        all_files="$all_files $arch/$file.sig"
-		elif test -n "$GPGKEY"
-		then
-			echo "Signing $arch/$file..." >&2
-			call_gpg --detach-sign --no-armor -u $GPGKEY "$dir/$arch/$file"
-			all_files="$all_files $arch/$file.sig"
-		fi
-	done
+			if test -f "$path".sig
+			then
+				cp "$path".sig "$dir/$arch/" ||
+				die "Could not copy $path.sig to $dir/$arch"
+	                        all_files="$all_files $arch/$file.sig"
+			elif test -n "$GPGKEY"
+			then
+				echo "Signing $arch/$file..." >&2
+				call_gpg --detach-sign --no-armor -u $GPGKEY "$dir/$arch/$file"
+				all_files="$all_files $arch/$file.sig"
+			fi
+		done
+	elif test "z${action}" = "zremove"
+	then
+		for pkgname in "$@"
+		do
+			mingw=
+			case "${pkgname}" in
+			mingw-w64-*)
+				arch=${pkgname##mingw-w64-}
+				arch=${arch#clang-}
+				arch=${arch%%-*}
+				key=${arch}_mingw
+				;;
+			git-extra)
+				# TODO we have no way to infer the arch from an msys pkgname
+				arch=i686
+				key=${arch}_mingw
+				;;
+			*)
+				# TODO we have no way to infer the arch from an msys pkgname
+				arch=i686
+				key=${arch}_msys
```

Comment:
I don't know what to do about the `arch` here... do I need to add another argument to specify which arch msys packages should be removed from?

Reply:
yeah, I just started looking at build-and-deploy.yml, and I think adding an architecture argument would be the way to go.